plugins/module_utils/firewall_rules_utils.py

```diff
@@ -0,0 +1,659 @@
+#  Copyright: (c) 2024, Puzzle ITC, Fabio Bertagna <bertagna@puzzle.ch>
+#  GNU General Public License v3.0+ (see LICENSE or https://www.gnu.org/licenses/gpl-3.0.txt)
+"""
+Utilities for firewall_rules module related operations.
+"""
+from dataclasses import dataclass, asdict, field
+from typing import List, Optional
+from xml.etree.ElementTree import Element
+
+from ansible_collections.puzzle.opnsense.plugins.module_utils import xml_utils
+from ansible_collections.puzzle.opnsense.plugins.module_utils.config_utils import (
+    OPNsenseModuleConfig,
+)
+
+from ansible_collections.puzzle.opnsense.plugins.module_utils.enum_utils import ListEnum
+
+
+# pylint: disable=too-few-public-methods
+class FirewallRuleAction(ListEnum):
+    """Represents the rule filter policy."""
+
+    PASS = "pass"
+    BLOCK = "block"
+    REJECT = "reject"
+
+
+# pylint: disable=too-few-public-methods
+class FirewallRuleDirection(ListEnum):
+    """Represents the rule direction."""
+
+    IN = "in"
+    OUT = "out"
+
+
+# pylint: disable=too-few-public-methods
+class FirewallRuleProtocol(ListEnum):
+    """Represents the protocol to filter in a rule."""
+
+    ANY = "any"
+    TCP = "tcp"
+    UDP = "udp"
+    TCP_UDP = "tcp/udp"
+    ICMP = "icmp"
+    ESP = "esp"
+    AH = "ah"
+    GRE = "gre"
+    IGMP = "igmp"
+    PIM = "pim"
+    OSPF = "ospf"
+    GGP = "ggp"
+    IPENCAP = "ipencap"
+    ST2 = "st2"
+    CBT = "cbt"
+    EGP = "egp"
+    IGP = "igp"
+    BBN_RCC = "bbn-rcc"
+    NVP = "nvp"
+    PUP = "pup"
+    ARGUS = "argus"
+    EMCON = "emcon"
+    XNET = "xnet"
+    CHAOS = "chaos"
+    MUX = "mux"
+    DCN = "dcn"
+    HMP = "hmp"
+    PRM = "prm"
+    XNS_IDP = "xns-idp"
+    TRUNK_1 = "trunk-1"
+    TRUNK_2 = "trunk-2"
+    LEAF_1 = "leaf-1"
+    LEAF_2 = "leaf-2"
+    RDP = "rdp"
+    IRTP = "irtp"
+    ISO_TP4 = "iso-tp4"
+    NETBLT = "netblt"
+    MFE_NSP = "mfe-nsp"
+    MERIT_INP = "merit-inp"
+    DCCP = "dccp"
+    PC = "3pc"
```

Comment:
Since variable names cannot start with integer values in python, i opted for this naming.
The molecule converge test, passes in this task, since operations and checks are done by the Enums value '3pc':
```
    - name: "Protocol: Test pass Rule with Protocol 3pc"
      puzzle.opnsense.firewall_rules:
        interface: 'lan'
        action: 'pass'
        protocol: '3pc'
        description: "New Test pass Rule with Protocol 3pc"
```